### PR TITLE
reissue: remove pipes from docker

### PR DESCRIPTION
**Pull request by minrk**
This DOES NOT need to get into the imminent 3.0 release.

removes use of pipes to resolve issues where builds could exit prematurely if test processes interacted with stdin/stdout (mpich's mpiexec may be the only known case).

This is a reissue of #337 with simplifying changes:

- use `docker run -e ENV_KEY` to pull env variables from the script env. This removes the need for an env file and separate create & start steps (thanks @isuruf!).
- use $CONFIG in DONE_CANARY and build.sh filenames to avoid collisions
- use more conservative `set -xeuo pipefail` for better error reporting of bugs in the script:
  - `-x` echoes commands
  - `-e` exits on first error
  - `-u` errors on undefined variables
  - `-o pipefail` makes exit code of piped commands make sense

See [this build](https://circleci.com/workflow-run/6aea75c0-2da5-4993-93a6-d280d9292c3f) for a successful build of a feedstock rerendered with this PR.

cf #695 which reverted #337

### CODE REVIEW

**Comment by isuruf** [comment]
@minrk, can you check that mpi and erlang works with this PR?

**Comment by minrk** [comment]
@isuruf will do. Pending mpich build at https://circleci.com/gh/minrk/mpich-feedstock/2

Feel free to push more to this PR if needed.

**Comment by minrk** [comment]
mpich [succeeded](https://circleci.com/gh/minrk/mpich-feedstock/2)

**Comment by isuruf** [comment]
Great. I checked zlib-feedstock as well. @jakirkham I'm fine with including this in 3.0.0. What do you think?

**Comment by jakirkham** [comment]
This looks promising.

Would be good to set the `BINSTAR_TOKEN` to some unique dummy value and see if it leaks through anywhere as a sanity check. Not seeing any obvious places where that would happen, but still would be nice as a safety measure.

**Comment by isuruf** [comment]
I checked whether `CONFIG` value is printed and it was printed only at the commands that used it.

**Comment by isuruf** [comment]
@minrk, can you fix the merge conflicts?

**Comment by isuruf** [comment]
cc @conda-forge/conda-smithy

**Comment by minrk** [comment]
@isuruf thanks for finishing this one!

**Comment by ocefpaf** [comment]
I guess this is good to go. Let's try it out!

**Comment by jakirkham** [comment]
Thanks for seeing this one through, @minrk. :)

File: conda_smithy/templates/run_docker_build.sh.tmpl [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+fi
+
+ARTIFACTS="$FEEDSTOCK_ROOT/build_artifacts"
+
+if [ -z "$CONFIG" ]; then
+    echo "Need to set CONFIG env variable"
+    exit 1
+fi
+
+test -d "$ARTIFACTS" || mkdir "$ARTIFACTS"
```
Comment (jakirkham) [inline conda_smithy/templates/run_docker_build.sh.tmpl]: Was there a reason we want to do this instead of `mkdir -p "$ARTIFACTS"`?

cc @pkgw

**Pull request merged**